src/Nonlinear/operators.jl

```diff
@@ -0,0 +1,748 @@
+    registered_univariate_operators::Vector{_UnivariateOperator}
+    # NODE_CALL_MULTIVARIATE
+    multivariate_operators::Vector{Symbol}
+    multivariate_operator_to_id::Dict{Symbol,Int}
+    multivariate_user_operator_start::Int
+    registered_multivariate_operators::Vector{_MultivariateOperator}
+    # NODE_LOGIC
+    logic_operators::Vector{Symbol}
+    logic_operator_to_id::Dict{Symbol,Int}
+    # NODE_COMPARISON
+    comparison_operators::Vector{Symbol}
+    comparison_operator_to_id::Dict{Symbol,Int}
+    function OperatorRegistry()
+        univariate_operators = copy(DEFAULT_UNIVARIATE_OPERATORS)
+        multivariate_operators = copy(DEFAULT_MULTIVARIATE_OPERATORS)
+        logic_operators = [:&&, :||]
+        comparison_operators = [:<=, :(==), :>=, :<, :>]
+        return new(
+            # NODE_CALL_UNIVARIATE
+            univariate_operators,
+            Dict{Symbol,Int}(
+                op => i for (i, op) in enumerate(univariate_operators)
+            ),
+            length(univariate_operators),
+            _UnivariateOperator[],
+            # NODE_CALL
+            multivariate_operators,
+            Dict{Symbol,Int}(
+                op => i for (i, op) in enumerate(multivariate_operators)
+            ),
+            length(multivariate_operators),
+            _MultivariateOperator[],
+            # NODE_LOGIC
+            logic_operators,
+            Dict{Symbol,Int}(op => i for (i, op) in enumerate(logic_operators)),
+            # NODE_COMPARISON
+            comparison_operators,
+            Dict{Symbol,Int}(
+                op => i for (i, op) in enumerate(comparison_operators)
+            ),
+        )
+    end
+end
+
+const _FORWARD_DIFF_METHOD_ERROR_HELPER = raw"""
+Common reasons for this include:
+
+ * the function assumes `Float64` will be passed as input, it must work for any
+   generic `Real` type.
+ * the function allocates temporary storage using `zeros(3)` or similar. This
+   defaults to `Float64`, so use `zeros(T, 3)` instead.
+
+As an example, instead of:
+```julia
+function my_function(x::Float64...)
+    y = zeros(length(x))
+    for i in 1:length(x)
+        y[i] = x[i]^2
+    end
+    return sum(y)
+end
+```
+use:
+```julia
+function my_function(x::T...) where {T<:Real}
+    y = zeros(T, length(x))
+    for i in 1:length(x)
+        y[i] = x[i]^2
+    end
+    return sum(y)
+end
+```
+
+Review the stacktrace below for more information, but it can often be hard to
+understand why and where your function is failing. You can also debug this
+outside of JuMP as follows:
+```julia
+import ForwardDiff
+
+# If the input dimension is 1
+x = 1.0
+my_function(a) = a^2
+ForwardDiff.derivative(my_function, x)
+
+# If the input dimension is more than 1
+x = [1.0, 2.0]
+my_function(a, b) = a^2 + b^2
+ForwardDiff.gradient(x -> my_function(x...), x)
+```
+"""
+
+_intercept_ForwardDiff_MethodError(err, ::Symbol) = rethrow(err)
+
+function _intercept_ForwardDiff_MethodError(::MethodError, op::Symbol)
```

Comment:
should the initial error message be pre/appended to this one?

Reply:
This happens automatically:
````julia
julia> using MathOptInterface; const MOI = MathOptInterface
MathOptInterface

julia> r = MOI.Nonlinear.OperatorRegistry()
MathOptInterface.Nonlinear.OperatorRegistry([:+, :-, :abs, :sqrt, :cbrt, :abs2, :inv, :log, :log10, :log2  …  :airyai, :airybi, :airyaiprime, :airybiprime, :besselj0, :besselj1, :bessely0, :bessely1, :erfcx, :dawson], Dict(:sinh => 39, :airyai => 63, :asind => 33, :sind => 21, :acot => 32, :acoth => 50, :asech => 48, :asinh => 45, :secd => 24, :erfinv => 54…), 72, MathOptInterface.Nonlinear._UnivariateOperator[], [:+, :-, :*, :^, :/, :ifelse], Dict(:+ => 1, :^ => 4, :/ => 5, :- => 2, :ifelse => 6, :* => 3), 6, MathOptInterface.Nonlinear._MultivariateOperator[], [:&&, :||], Dict(:|| => 2, :&& => 1), [:<=, :(==), :>=, :<, :>], Dict(:(==) => 2, :>= => 3, :> => 5, :<= => 1, :< => 4))

julia> f(x::Float64) = x^2
f (generic function with 1 method)

julia> MOI.Nonlinear.register_operator(r, :f, 1, f)
ERROR: Unable to register the function :f because it does not support differentiation via ForwardDiff.

Common reasons for this include:

 * the function assumes `Float64` will be passed as input, it must work for any
   generic `Real` type.
 * the function allocates temporary storage using `zeros(3)` or similar. This
   defaults to `Float64`, so use `zeros(T, 3)` instead.

As an example, instead of:
```julia
function my_function(x::Float64...)
    y = zeros(length(x))
    for i in 1:length(x)
        y[i] = x[i]^2
    end
    return sum(y)
end
```
use:
```julia
function my_function(x::T...) where {T<:Real}
    y = zeros(T, length(x))
    for i in 1:length(x)
        y[i] = x[i]^2
    end
    return sum(y)
end
```

Review the stacktrace below for more information, but it can often be hard to
understand why and where your function is failing. You can also debug this
outside of JuMP as follows:
```julia
import ForwardDiff

# If the input dimension is 1
x = 1.0
my_function(a) = a^2
ForwardDiff.derivative(my_function, x)

# If the input dimension is more than 1
x = [1.0, 2.0]
my_function(a, b) = a^2 + b^2
ForwardDiff.gradient(x -> my_function(x...), x)
```

Stacktrace:
 [1] error(s::String)
   @ Base ./error.jl:33
 [2] _validate_register_assumptions(f::typeof(f), name::Symbol, dimension::Int64)
   @ MathOptInterface.Nonlinear ~/.julia/dev/MathOptInterface/src/Nonlinear/operators.jl:249
 [3] MathOptInterface.Nonlinear._UnivariateOperator(op::Symbol, f::Function)
   @ MathOptInterface.Nonlinear ~/.julia/dev/MathOptInterface/src/Nonlinear/operators.jl:263
 [4] register_operator(::MathOptInterface.Nonlinear.OperatorRegistry, ::Symbol, ::Int64, ::Function, ::Vararg{Function, N} where N)
   @ MathOptInterface.Nonlinear ~/.julia/dev/MathOptInterface/src/Nonlinear/operators.jl:322
 [5] top-level scope
   @ REPL[6]:1

caused by: MethodError: no method matching f(::ForwardDiff.Dual{ForwardDiff.Tag{typeof(f), Float64}, Float64, 1})
Closest candidates are:
  f(::Float64) at REPL[5]:1
Stacktrace:
 [1] derivative
   @ ~/.julia/packages/ForwardDiff/PBzup/src/derivative.jl:14 [inlined]
 [2] _validate_register_assumptions(f::typeof(f), name::Symbol, dimension::Int64)
   @ MathOptInterface.Nonlinear ~/.julia/dev/MathOptInterface/src/Nonlinear/operators.jl:243
 [3] MathOptInterface.Nonlinear._UnivariateOperator(op::Symbol, f::Function)
   @ MathOptInterface.Nonlinear ~/.julia/dev/MathOptInterface/src/Nonlinear/operators.jl:263
 [4] register_operator(::MathOptInterface.Nonlinear.OperatorRegistry, ::Symbol, ::Int64, ::Function, ::Vararg{Function, N} where N)
   @ MathOptInterface.Nonlinear ~/.julia/dev/MathOptInterface/src/Nonlinear/operators.jl:322
 [5] top-level scope
   @ REPL[6]:1
````